haskell-overlays/ghcjs.nix

```diff
@@ -1,8 +1,18 @@
-{ lib, haskellLib, nixpkgs, fetchgit, fetchFromGitHub, useReflexOptimizer }:
+{ system, lib, haskellLib, nixpkgs, fetchgit, fetchFromGitHub, useReflexOptimizer }:
 
 with haskellLib;
 
-self: super: {
+self: super:
+
+let dontHardcodeLinux = package: cabalFile:
+  if ! self.ghc.isGhcjs then package
+  else nixpkgs.haskell.lib.overrideCabal package (drv: {
+    postPatch = (drv.postPatch or "") + nixpkgs.lib.optionalString (system == "x86_64-darwin") ''
+      substituteInPlace ${cabalFile}.cabal --replace 'if os(linux)' 'if os(linux) && !impl(ghcjs)'
+      substituteInPlace ${cabalFile}.cabal --replace 'if os(osx)' 'if os(linux) && impl(ghcjs)'
```

Comment:
Is this right? `if os(osx)` gets translated to `if os(linux) ...`?

Reply:
If it is right...an explanation is worthy.

Reply:
Talked to @alexfmpe. So basically this is just a very confusing way of writing it. Because of the hardcoding in GHCJS, `os(linux) == true` and `true && x == x` so why are we confounding this substitution?

Reply:
We don't want the `os(linux)` patches to get applied, so we change the `os(linux)` conditional to only apply when GHCJS isn't being used. Since GHCJS always thinks it's running on Linux, we change the `os(osx)` patches to `os(linux)` patches.

Reply:
see https://github.com/kadena-io/pact/commit/98e6e365c3a5dc508579b7af56d8e924dd6ed86f

Reply:
But if GHCJS is hardcoding then why not just drop `os(linux)` completely? It's always `true`. It seems to just confuse the matter.

Reply:
Sure, I guess we could do that.

Reply:
Also I'm a bit confused why the os(osx) case requires impl(ghcjs) while the os(linux) one requires *not* impl(ghcjs).

Reply:
because we don't want to apply the *actual* `os(linux)` patches (since we're really on macOS)

Reply:
GHCJS thinks we're on Linux, but that doesn't mean we should apply the Linux patches, instead we need to trick GHCJS into applying the macOS patches

Reply:
so we say "these linux patches are not for you, silly GHCJS, but these other (actually macOS) linux patches are!"

Reply:
Note the replacement only happens if `system == "x86_64-darwin"`

Reply:
That makes sense. Thanks.